### ISSUE TITLE
Failed to create database "metastore_db" using spark-jobserver 0.8.1

### ISSUE THREAD

**Issue by barrybecker4** [spark-jobserver/spark-jobserver#1064]
Spark 2.3.0
latest 0.8.1 from master (~4/20/2018).
"client" Deployed mode
Red Hat enterprise Linux

I am upgrading my server to spark 2.3.0 and job-server 0.8.1 from spark 2.1.1 and job-server 0.8.0 (which were working fine). I am using the JobSqlDao with MySql and am using the SessionContextFactory to create a sqlContext. In local.conf, I have:
```
 sql-context {
      context-factory = spark.jobserver.context.SessionContextFactory
 }
```
Everything started up fine, and everything looks good at ports 8090 and 4040, but when I tried to issue my first request to job-server, I got the below error. From searching the web, this looks like something to do with Hive, but I'm not sure. I don't use hive, but maybe the SessionContextFactory needs it. Any thoughts? 
```
java.sql.SQLException: Unable to open a test connection to the given database. JDBC url = jdbc:derby:;databaseName=metastore_db;create=true, username = APP. Terminating connection pool (set lazyInit to true if you expect to start your database after your app). Original Exception: ------
java.sql.SQLException: Failed to create database 'metastore_db', see the next exception for details.
:
Caused by: ERROR XJ041: Failed to create database 'metastore_db', see the next exception for details.
        at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
        at org.apache.derby.impl.jdbc.SQLExceptionFactory.wrapArgsForTransportAcrossDRDA(Unknown Source)
        ... 105 more
Caused by: ERROR XBM0H: Directory /metastore_db cannot be created.
        at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
        at org.apache.derby.iapi.error.StandardException.newException(Unknown Source)
        at org.apache.derby.impl.services.monitor.StorageFactoryService$10.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.apache.derby.impl.services.monitor.StorageFactoryService.createServiceRoot(Unknown Source)
        at org.apache.derby.impl.services.monitor.BaseMonitor.bootService(Unknown Source)
        at org.apache.derby.impl.services.monitor.BaseMonitor.createPersistentService(Unknown Source)
        at org.apache.derby.impl.services.monitor.FileMonitor.createPersistentService(Unknown Source)
        at org.apache.derby.iapi.services.monitor.Monitor.createPersistentService(Unknown Source)
        at org.apache.derby.impl.jdbc.EmbedConnection$5.run(Unknown Source)
        at java.security.AccessController.doPrivileged(Native Method)
        at org.apache.derby.impl.jdbc.EmbedConnection.createPersistentService(Unknown Source)
```

**Comment by barrybecker4** [comment]
I added a hive-site.xml file to spark/conf which points the metastore_db to a place that it can write to. The contents of my hive-site.xml file is:
```
<?xml version="1.0"?>
<?xml-stylesheet type="text/xsl" href="configuration.xsl"?>
<configuration>
<property>
  <name>javax.jdo.option.ConnectionURL</name>
  <value>jdbc:derby:;databaseName=/home/mineset/metastore_db;create=true</value>
  <description>JDBC connect string for a JDBC metastore</description>
</property>
<property>
  <name>javax.jdo.option.ConnectionDriverName</name>
  <value>org.apache.derby.jdbc.EmbeddedDriver</value>
</property>
</configuration>
```
but now I get a different error:
```
Caused by: java.net.ConnectException: Call From ra.esi-internal.esi-group.com/192.168.xxx.xx to 192.168.xxx.xx:8020 failed on connection exception: java.net.ConnectException: Connection refused; For more details see:  http://wiki.apache.org/hadoop/ConnectionRefused
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
        at org.apache.hadoop.net.NetUtils.wrapWithMessage(NetUtils.java:791)
```
Is there a context factory that I can use to get a sqlContext that does not require hive? This SessionContextFactory is new in 0.8.1, and seems to be the source of my problems.

**Comment by barrybecker4** [comment]
Since I'm desperate, I also posted to stack-overflow:
https://stackoverflow.com/questions/50108069/failed-to-create-database-metastore-db-using-spark-jobserver-0-8-1

**Comment by noorul** [comment]
@barrybecker4 I am not sure whether SJS is qualified to run along with spark 2.3.0

**Comment by barrybecker4** [comment]
Perhaps not, but we need to use spark 2.3 for other reasons and need the latest spark job-server to avoid other problems that noticed when using the JobSqlDao with earlier versions of sjs. 

I found a workaround for my hive issues. If found that if I remove all the hive jars from the prebuilt spark distribution, then it sees that there is no spark support and uses "in-memory" instead of "hive" for the calalogImplementation. I wish I could just configure it to not use hive, but I could not find a way. I tried configuring it to use in-memory for spark.sql.catalogImplementation, but it did not take if the hive jars were present. Here are the list of jars that I removed:

hive-beeline-1.2.1.spark2.jar
hive-exec-1.2.1.spark2.jar
hive-metastore-1.2.1.spark2.jar
spark-hive-thriftserver_2.11-2.3.0.jar
hive-cli-1.2.1.spark2.jar
hive-jdbc-1.2.1.spark2.jar
spark-hive_2.11-2.3.0.jar

**Comment by thoHeinze** [comment]
@barrybecker4 I think you problem is https://github.com/spark-jobserver/spark-jobserver/blob/master/job-server-extras/src/main/scala/spark/jobserver/context/SessionContextFactory.scala#L26

We should add a jobserver parameter here to allow the user to choose if he wants to use Hive or not. Currently it assume you always to use hive as long as the Hive dependencies are in your path. By removing them like you did, you will hit the catch block in line 28.

**Comment by barrybecker4** [comment]
It is running now, but there are problems. How long do you think it will be before jobserver 0.8.1 is qualified for spark 2.3.0?
We see slow performance and occasional crashes where the sql-context is destroyed with an error message like
> Message [spark.jobserver.io.JobDAOActor$LastUploadTimeAndType] from Actor[akka://JobServer/user/dao-manager#273012949] to Actor[akka://JobServer/deadLetters] was not delivered. [5] dead letters encountered. This logging can be turned off or adjusted with configuration settings 'akka.log-dead-letters' and 'akka.log-dead-letters-during-shutdown'.
> [2018-05-01 12:17:50,678] ERROR .jobserver.JobManagerActor [] [] - About to restart actor due to exception:
> java.util.concurrent.TimeoutException: Futures timed out after [3 seconds]

I increased daoAskTimeout from 3 seconds to 15 seconds in JobManagerActor and recompiled, but that seems like only fixing the symptom of some deeper problem. 

Not sure what explains the slower performance. We are now using JobSqlDao with MySql (were using JobFileDao before). The jar that we upload to jobserver is fairly large - 46m. Could using dependent-lib-uri parameter to split out dependent jars have an impact on performance? The other thing I noticed is that there are many copies of the uploaded jar being put in /tmp/spark-jobserver/sqlDao/data and in the database. Is writing all these large jar files to the db an issue? I'm not sure if there are deploy modes other than "client" or "server" that I should consider. Here is my local.conf file:
```
spark {
  # spark.master will be passed to each job's JobContext
  master = "spark://192.168.xxx.xx:7077"

  # Default # of CPUs for jobs to use for Spark standalone cluster
  job-number-cpus = 4

  # Client or cluster deployment. See https://github.com/spark-jobserver/spark-jobserver/issues/720
  submit.deployMode = "client"

  jobserver {
    port = 8090
    context-per-jvm = false
    max-jobs-per-context = 512
    context-creation-timeout = 20s
    short-timeout = 60s   # needed for fat jars

    // Using spark_jobserver db in mysql to record logging
    jobdao = spark.jobserver.io.JobSqlDAO

    filedao {
      rootdir = /home/mineset/tmp/spark-jobserver/filedao/data
    }

    datadao {
      # storage directory for files that are uploaded to the server
      # via POST/data commands
      rootdir = /home/mineset/tmp/spark-jobserver/upload
    }

    job-jar-paths {
      mineset-spark = /home/mineset/tmp/spark-jobserver/jars/mineset-spark.jar
    }

    sqldao {
      slick-driver = slick.driver.MySQLDriver
      jdbc-driver = com.mysql.jdbc.Driver
      rootdir = /home/mineset/tmp/spark-jobserver/sqldao/data
      jdbc {
        url = "jdbc:mysql://localhost/spark_jobserver?createDatabaseIfNotExist=true&useSSL=false"
        user = "jobserver"
        password = "spark"
      }

      # DB connection pool settings
      dbcp {
        enabled = false
        maxactive = 20
        maxidle = 10
        initialsize = 10
      }
    }
    # When using chunked transfer encoding with scala Stream job results, this is the size of each chunk
    result-chunk-size = 1m
  }

  # Predefined Spark Contexts
  contexts {
    sql-context {
      context-factory = spark.jobserver.context.SessionContextFactory
    }
  }

  # Universal context configuration.  These settings can be overridden, see README.md
  context-settings {
    spark.ui.port = 4040
    spark.shuffle.manager = sort
    spark.executor.memory = 20G
    spark.sql.shuffle.partitions = 48
    spark.cores.max = 24
    spark.driver.maxResultSize = 2G
    spark.scheduler.mode = "FAIR"
    spark.worker.cleanup.enabled = true
    spark.worker.cleanup.appDataTtl = 36000
    spark.worker.cleanup.interval = 1800
    spark.serializer = org.apache.spark.serializer.KryoSerializer
    spark.kryo.classesToRegister = scala.collection.mutable.LinkedHashMap
    passthrough {
      #es.nodes = "192.1.1.1"
    }
  }
}

spray.can.server {
  idle-timeout = 360 s
  request-timeout = 300 s
  parsing.max-content-length = 400m
}
```

**Comment by barrybecker4** [comment]
Increasing the daoAskTimeout in the code seems to have worked around the problem with the crashing for now. Performance seems comparable to before, but we still need to do more testing. Large requests cause https://github.com/spark-jobserver/spark-jobserver/issues/1065.

**Comment by barrybecker4** [comment]
adding another note here for whoever might be trying to run sjs 0.8.1 with spark 2.3.0.
We run a spark cluster in standalone mode. For the longest time we could not understand why the executors for the workers were not showing up in the executors tab. Upon comparing the env tab in the new install with the env tab in the old version, I noticed that master was set to local[*] instead of spark://192.168.10.1:7077.  As shown above, we have master = "spark://192.168.xxx.xx:7077" set in local config, but it seems that is lo longer honored! To fix, I had to move this setting from the outside spark block into the context-settings block and call it 
  spark.master="spark://192.168.xxx.xx:7077"
Then it worked. Is this expected based on some change in 0.8.1 or is it a bug?
